Slash Commands and Interactions

msciotti opened the pull request:
Opening the PR now because the "meat" of the docs are done. Just a quick note:

Comments on this PR should be solely about the **documentation**. When this launches, we will have specific repo templates for feedback on the feature itself. I will pretty aggressively prune the convo here to be documentation-only so that we can get this out

This documentation page is a bit of a monolith, and should probably be split up into smaller sections. The same can be said of most of our documentation, to be honest. I'm prioritizing having the documentation ready over having it look pretty, and we can revisit later.

TODO:

- [x] Write a better intro
- [x] Document `INTERACTION_CREATE` gateway event
- [x] Document subcommands and groups
- [x] How does the security check actually work
- [x] Make sure all the links work
- [x] Put a resources section since we have a nice node and python library to help with cryptography and some data models
- [x] Check styling in local dev for 🌟 aesthetics 🌟 
- [x] Make sure the code examples actually work
- [x] Document limits

msciotti committed:
Slash Commands and Interactions documentation

msciotti committed:
Nits and TODOs

msciotti committed:
Merge branch 'master' into feature/interactions

msciotti committed:
create, not created

yonilerner commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,470 @@
+
+If you are receiving Interactions via outgoing webhook, there are some security steps you **must** take before your app is eligible to receive requests. The internet is a scary place, especially for people hosting open, unauthenticated endpoints.
+
+Every Interaction is sent with a `x-signature-ed25519` header. The header includes an encryption key and a timestamp. Using your favorite security library of choice, you **must validate this signature header each time to you receive an interaction**. If the signature fails validate, respond with a `400` error code.
```
Does this have to be 400 or can it be 401?

msciotti replied:
Good question--I'll ask night this morning.

msciotti replied:
Funny enough, we actually are fine with anything not-2XX. But we'll change this to say 401 for API cleanliness.

msciotti committed:
Fix python make command example

msciotti committed:
More endpoints

msciotti committed:
groups and subcommands

yonilerner commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,706 @@
+
+If you are receiving Interactions via outgoing webhook, there are some security steps you **must** take before your app is eligible to receive requests. The internet is a scary place, especially for people hosting open, unauthenticated endpoints.
+
+Every Interaction is sent with an `x-signature-ed25519` header. The header includes an encryption key and a timestamp. Using your favorite security library of choice, you **must validate this signature header each time to you receive an interaction**. If the signature fails validate, respond with a `401` error code.
```
I think we should elaborate here on what exactly is required. eg.
> We send `x-signature-timestamp` and `x-signature-ed255191`. Concatenate the timestamp with the request body and validate that message against the signature

msciotti replied:
The newest docs have some sample code in there now!

msciotti committed:
should be application id

msciotti committed:
interaction create event

msciotti committed:
Rate limits and maximums

msciotti committed:
http verbs

msciotti committed:
Signature validation examples

msciotti committed:
More limits

night commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,814 @@
+
+ const signature = req.get('X-Signature-Ed25519');
+ const timestamp = req.get('X-Signature-Timestamp');
+ const isValidRequest = await verifyKey(req.rawBody, signature, timestamp, 'MY_CLIENT_PUBLIC_KEY');
```
let's recommend tweetnacl since it's more popular and has been audited.

```
const nacl = require('tweetnacl');

const PUBLIC_KEY = 'APPLICATION_PUBLIC_KEY';

const signature = req.get('X-Signature-Ed25519');
const timestamp = req.get('X-Signature-Timestamp');
const body = req.rawBody; // rawBody is expected to be a string, not raw bytes

try {
   const isVerified = nacl.sign.detached.verify(
    Buffer.from(timestamp + body),
    Buffer.from(signature, 'hex'),
    Buffer.from(PUBLIC_KEY, 'hex')
  );
  if (!isVerified) {
    throw new Error('invalid signature');
  }
} catch (error) {
  return res.status(401).end(`invalid request signature: ${error.message}`);
}
```

night commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,814 @@
+signature = request.headers["X-Signature-Ed25519"]
+timestamp = request.headers["X-Signature-Timestamp"]
+
+if not verify_key(request.data, signature, timestamp, 'my_client_public_key'):
```
let's update this example to demonstrate raw pynacl usage:

```
from nacl.signing import VerifyKey
from nacl.exceptions import BadSignatureError

PUBLIC_KEY = 'APPLICATION_PUBLIC_KEY'

verify_key = VerifyKey(bytes.fromhex(PUBLIC_KEY))

signature = request.headers["X-Signature-Ed25519"]
timestamp = request.headers["X-Signature-Timestamp"]
body = request.data

try:
    verify_key.verify(f'{timestamp}{body}'.encode(), bytes.fromhex(signature))
except BadSignatureError:
    abort(401, 'invalid request signature')
```

msciotti committed:
a better intro

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+url = "https://discord.com/api/v8/interactions/<interaction_id>/<interaction_token>/callback"
+
+json = {
+    "type": 1,
```
```suggestion
    "type": 4,
```

my bad; type 1 doesn't make sense here

AEnterprise commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+                        {
+                            "name": "role",
+                            "description": "The role to get",
+                            "type": 6, // 8 is type ROLE
```
sending type 6 when the comment and name are for type 8 role

AEnterprise commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+                        {
+                            "name": "role",
+                            "description": "The role to edit",
+                            "type": 6, // 8 is type ROLE
```
same thing here

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+> danger
+> Creating a command with the same name as an existing command for your application will overwrite the old command.
+
+Create a new global commands. New global commands will be available in all guilds after 1 hour. Returns `200` and an [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) object.
```
```suggestion
Create a new guild-specific command. Returns `200` and an [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) object.
```

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+
+## Edit Guild Application Command % PATCH /applications/{application.id}/guilds/{guild.id#DOCS_RESOURCES_GUILD/guild-object}/commands
+
+Edit a global commands. Updates will be available in all guilds after 1 hour. Returns `200` and an [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) object.
```
```suggestion
Edit a guild-specific command. Returns `200` and an [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) object.
```

Zomatree replied:
same here, this shouldnt say "global commands", and the updates shouldnt take 1 hour either.

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+
+## Delete Guild Application Command % DELETE /applications/{application.id}/commands/{id}
+
+Deletes a guild commands. Returns `204` on success.
```
```suggestion
Deletes a guild-specific command. Returns `204` on success.
```

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+> danger
+> Creating a command with the same name as an existing command for your application will overwrite the old command.
+
+Create a new global commands. New global commands will be available in all guilds after 1 hour. Returns `200` and an [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) object.
```
```suggestion
Create a new global command. New global commands will be available in all guilds after 1 hour. Returns `200` and an [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) object.
```

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+
+## Edit Global Application Command % PATCH /applications/{application.id}/commands/{id}
+
+Edit a global commands. Updates will be available in all guilds after 1 hour. Returns `200` and an [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) object.
```
```suggestion
Edit a global command. Updates will be available in all guilds after 1 hour. Returns `200` and an [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) object.
```

Zomatree commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+
+## Get Guild Application Commands % GET /applications/{application.id}/guilds/{guild.id#DOCS_RESOURCES_GUILD/guild-object}/commands
+
+Fetch all of the global commands for your application. Returns a list of [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) objects.
```
This shouldnt say "global commands" as its the guild commands

advaith1 replied:
```suggestion
Fetch all of the global commands for your application. Returns an array of [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) objects.
```

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+
+## Delete Global Application Command % DELETE /applications/{application.id}/commands/{id}
+
+Deletes a global commands. Returns `204`.
```
```suggestion
Deletes a global command. Returns `204`.
```

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+
+## Get Global Application Commands % GET /applications/{application.id}/commands
+
+Fetch all of the global commands for your application. Returns a list of [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) objects.
```
```suggestion
Fetch all of the global commands for your application. Returns an array of [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command) objects.
```

Zomatree commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+| description | string                                                                                            | 1-100 character description    |
+| options?    | array of [ApplicationCommandOption](#DOCS_INTERACTIONS_SLASH_COMMANDS/application-command-option) | the parameters for the command |
+
+## Edit Guild Application Command % PATCH /applications/{application.id}/guilds/{guild.id#DOCS_RESOURCES_GUILD/guild-object}/commands
```
shouldnt this have `/{id}` at the end?

AEnterprise reviewed (commented):
there is also a new message type for when a user uses a command, can this be documented as well?

AEnterprise commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,820 @@
+
+In this documentation you'll find some notes about limits and caps on certain parts of Slash Commands. At a high level, they are as follows:
+
+- An app can have up to 50 global commands
```
what counts towards this limit? 50 top level commands or, do groups and subcommands count towards this as well?

msciotti replied:
50 top-level commands. I think there's a max of subcommands and groups that should be documented as well

msciotti committed:
swap code samples for sig validation

msciotti committed:
global -> guild, list -> array

msciotti committed:
Fix missing id params

msciotti committed:
wrong role type

msciotti committed:
NOW the right role type

BrightSkyz commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,837 @@
+            "discriminator": "1337",
+            "public_flags": 131141
+        },
+        "roles": [539082325061836999],
```
Shouldn't this role snowflake be a string? When I looked at the request body I got when executing an interaction, the snowflake was a string.

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,837 @@
+> info
+> For authoriation, all endpoints take either a bot token or client credentials token for your application
+
+## Get Global Application Commands % GET /applications/{application.id}/commands
```
`{application.id}` should probably link to the application object; would be good to merge #1838 first so its in a better place

also can application data in READY be documented? would be useful for libs/old bots implementing this feature

msciotti replied:
Funny enough I think we only document the application object as a response to the GET endpoint. Probably need to refactor that a bit

advaith1 replied:
the interaction response & followup message endpoints weren't linked, looks like they don't have the {}

webtax-gh commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,837 @@
+|---------|--------------------------------------------------|-----------------------------------|
+| id      | snowflake                                        | the ID of the invoked command     |
+| name    | string                                           | the name of the invoked command   |
+| options | array of ApplicationCommandInteractionDataOption | the params + values from the user |
```
Shouldn't it be like this?

```suggestion
| options? | array of ApplicationCommandInteractionDataOption | the params + values from the user |
```

msciotti committed:
PR Fixes and Application Object
- Added application object
- roles are a string
- applications.commands.update scope

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,837 @@
+
+Edits the initial Interaction response. Functions the same as [Edit Webhook Message](#DOCS_RESOURCES_WEBHOOK/edit-webhook-messages).
+
+# Delete Original Interaction Response % DELETE /webhooks/application.id/{interaction.token#DOCS_INTERACTIONS_SLASH_COMMANDS/interaction}/messages/@original
```
both the delete headers are level 1 instead of level 2 
```suggestion
## Delete Original Interaction Response % DELETE /webhooks/application.id/{interaction.token#DOCS_INTERACTIONS_SLASH_COMMANDS/interaction}/messages/@original
```

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,837 @@
+
+Edits a followup message for an Interaction. Functions the same as [Edit Webhook Message](#DOCS_RESOURCES_WEBHOOK/edit-webhook-messages).
+
+# Delete Followup Message % DELETE /webhooks/application.id/{interaction.token#DOCS_INTERACTIONS_SLASH_COMMANDS/interaction}/messages/{message.id#DOCS_RESOURCES_CHANNEL/message-object}
```
```suggestion
## Delete Followup Message % DELETE /webhooks/application.id/{interaction.token#DOCS_INTERACTIONS_SLASH_COMMANDS/interaction}/messages/{message.id#DOCS_RESOURCES_CHANNEL/message-object}
```

msciotti committed:
application on ready

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,837 @@
+If your application does not require a bot user within the guild for its commands to work, **you no longer need to add for the bot scope or specific permissions**. To clarify this point even more:
+
+> danger
+> If your application only talks to Discord through creating Slash Commands and responding to Interactions, and does not use any other part of our API, **you no longer need to request the bot scope**.
```
why is this a danger? an info is probably better

```suggestion
> info
> If your application only talks to Discord through creating Slash Commands and responding to Interactions, and does not use any other part of our API, **you no longer need to request the bot scope**.
```

msciotti committed:
headings

FasterSpeeding commented on docs/topics/OAuth2.md:
```diff
@@ -386,6 +387,7 @@ Returns the bot's OAuth2 application info.
 | primary_sku_id?        | snowflake                                                  | if this application is a game sold on Discord, this field will be the id of the "Game SKU" that is created, if exists     |
 | slug?                  | string                                                     | if this application is a game sold on Discord, this field will be the URL slug that links to the store page               |
 | cover_image?           | string                                                     | if this application is a game sold on Discord, this field will be the hash of the image on store embeds                   |
+| flags                  | int                                                        | the application's public flags                                                                                            |
```
Should flags not be marked as it's not returned by `/oauth2/applications/@me`?

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,837 @@
+
+## Receiving an Interaction
+
+When a user completes a Slash Commands, your app will receive an **Interaction**. Your app can receive an interaction in one of two ways:
```
```suggestion
When a user uses a Slash Command, your app will receive an **Interaction**. Your app can receive an interaction in one of two ways:
```

msciotti committed:
10s all the way down

ThaTiemsz commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,837 @@
+
+If you are receiving Interactions over the gateway, you will **also need to respond via HTTP**. Responses to Interactions **are not sent as commands over the gateway**.
+
+To respond to a gateway Interaction, make a `POST` request like this. `interaction_id` is the unique id of that individual Interaction from the receieved payload. `interaction_token` is the unique token for that interaction from the received payload. **This endpoint is only valid for Interactions received over the gateway. Otherwise, respond to the `POST` request.**
```
```suggestion
To respond to a gateway Interaction, make a `POST` request like this. `interaction_id` is the unique id of that individual Interaction from the received payload. `interaction_token` is the unique token for that interaction from the received payload. **This endpoint is only valid for Interactions received over the gateway. Otherwise, respond to the `POST` request.**
```

msciotti committed:
grammar and spelling

msciotti committed:
broken links

msciotti committed:
more broken links

msciotti committed:
what even?

msciotti committed:
options -> object

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,839 @@
+
+## Create Interaction Response % POST /interactions/{interaction.id#DOCS_INTERACTIONS_SLASH_COMMANDS/interaction}/{interaction.token#DOCS_INTERACTIONS_SLASH_COMMANDS/interaction}/callback
+
+Create a response to an Interaction from the gateway. Functions the same as [Execute Webhook](#DOCS_RESOURCES_WEBHOOK/execute-webhook)
```
this does **not** function the same as execute webhook, should link to the Interaction Response section at the bottom

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,839 @@
+| content           | string           | message content                                                                             |
+| embeds?           | array of embeds  | supports up to 10 embeds                                                                    |
+| allowed_mentions? | allowed mentions | [allowed mentions](#DOCS_RESOURCES_CHANNEL/allowed-mentions-object) object                  |
+| flags             | int              | acceptable values are [message flags](#DOCS_RESOURCES_CHANNEL/message-object-message-flags) |
```
this can only be used to set the EPHEMERAL flag; probably shouldn't document it at all if ephemeral messages arent being documented

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,839 @@
+## ApplicationCommand
+
+> danger
+> A command, or each individual subcommand, can have a maximum of 10 `options`
```
these are also better as info imo

```suggestion
> info
> A command, or each individual subcommand, can have a maximum of 10 `options`
```

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,839 @@
+## ApplicationCommandOption
+
+> danger
+> You can specify a maximum of 10 `choices` per option
```
```suggestion
> info
> You can specify a maximum of 10 `choices` per option
```

advaith1 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,839 @@
+- `PATCH /webhooks/<application_id>/<interaction_token>/messages/<message_id>` to edit a message sent with that `token`
+
+> danger
+> Interactions webhooks share the same rate limit properties as normal webhooks
```
maybe give some info about this?

msciotti committed:
one more broken link

msciotti committed:
make things less dangerous

advaith1 commented on docs/resources/Webhook.md:
```diff
@@ -152,6 +152,8 @@ Add a new webhook to your GitHub repo (in the repo's settings), and use this end
 
 ## Edit Webhook Message % PATCH /webhooks/{webhook.id#DOCS_RESOURCES_WEBHOOK/webhook-object}/{webhook.token#DOCS_RESOURCES_WEBHOOK/webhook-object}/messages/{message.id#DOCS_RESOURCES_CHANNEL/message-object}
 
+Edits a previously-sent webhook message from the same token. Takes the same JSON parameters as [Execute Webhook](#DOCS_RESOURCES_WEBHOOK/execute-webhook).
```
it doesn't take all the params execute webhook does; the params it does take are listed below so this isn't needed

msciotti replied:
Think this got messed up in the merge.

msciotti committed:
not webhook

advaith1 commented on docs/Change_Log.md:
```diff
@@ -1,5 +1,16 @@
+
+Slash Commands include some new features for webhooks as well:
+
+- Webhooks can now update previously-sent message from the same webhook using [Edit Webhook Message](#DOCS_RESOURCES_WEBHOOK/edit-webhook-message)
```
```suggestion
- Webhooks can now update previously-sent messages from the same webhook using [Edit Webhook Message](#DOCS_RESOURCES_WEBHOOK/edit-webhook-message) and [Delete Webhook Message](#DOCS_RESOURCES_WEBHOOK/delete-webhook-message)
```

msciotti committed:
Broken links and spelling

Co-authored-by: Advaith <advaithj1@gmail.com>

msciotti committed:
fix old merge issue in webhook edit

msciotti committed:
subcommand warning

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+Slash Commands are the new, exciting way to build and interact with apps on Discord.
+
+With Slash Commands, all you have to do is type `/` and you're ready to use your favorite bot. Users can learn everything your bot does and easily find new features as you add them. Validation, error states, and helpful UI walks them through your commands, meaning they can get it right the first time, especially on mobile. You now have one more ally in the fight against your phone's autocorrect.
```
> Validation

This is probably feature-level, but will there be something where you can for example return a 400 status code if the arguments are wrong as a sort of lower-level validation? Type-based validation doesn't roll by when you have for example a bot that rolls a six sides and you have to choose one. (ignore how this could just be a choice argument in the current implementation)

msciotti replied:
We don't have a way for passing non-200 back to the user because we assume it's your API telling Discord that something failed, not necessarily bad user input. It's a bit counterintuitive but we generally expect you to return 200 w/ an error to a user.

Definitely feature request-able and can be re-written as such tomorrow.

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+A **Slash Command** is a command that you register for your application. They're made up of a name, description, and a block of `options`, which you can think of like arguments to a function. The name and description help users find your command among many others, and the `options` validate user input as they fill out your command.
+
+Your commands are available in every guild that adds your application. You can also make commands for a specific guild; they're only available in that guild.
```
Might want to specify only global commands will be available to every guild, even if it is sort of specified right after.

A5rocks commented on docs/interactions/Slash_Commands.md:
````diff
@@ -0,0 +1,841 @@
+        })
+```
+
+You'll also need to properly set up [Security and Authorization](#DOCS_INTERACTIONS_SLASH_COMMANDS/security-and-authorization) on your endpoint for the URL to be accepted. Once both of those are complete and your URL has been saved, you can start receiving Interactions via webhook! At this point, your app will **no longer receive Interactions over the gateway**. If you want to receive them over t ...
````
Do you receive them _immediately_ or after you restart your bot's gateway?

msciotti replied:
Immediately

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+Interactions--both receiving and responding--are webhooks under the hood. So responding to an Interaction is just like sending a webhook request!
+
+When responding to an interaction received **via webhook**, your server can simply respond to the received `POST` request. You'll want to respond with a `200` status code (if everything went well), as well as specifying a `type` and `data`, which is an object with parameters the same as [Execute Webhook](#DOCS_RESOURCES_WEBHOOK/execute-webhook):
```
As @advaith1 has pointed out for another section, this should probably point to the Interaction Response data type at the bottom, which is mentions "Not all message fields are currently supported."

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+If you are receiving Interactions over the gateway, you will **also need to respond via HTTP**. Responses to Interactions **are not sent as commands over the gateway**.
+
+To respond to a gateway Interaction, make a `POST` request like this. `interaction_id` is the unique id of that individual Interaction from the received payload. `interaction_token` is the unique token for that interaction from the received payload. **This endpoint is only valid for Interactions received over the gateway. Otherwise, respond to the `POST` request.**
```
I'm not sure about the wording on this, as it makes it seem like you cannot use REST to send a message when using the outgoing webhook - you can though, seeing that "Create Followup Message" and whatnot are a thing.

In specific:
> Otherwise, respond to the `POST` request.

msciotti replied:
I added a bit to the start of the followup messages section that I think helps clarify this.

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+Interaction tokens are valid for **15 minutes**, meaning you can respond and update to a Slash Command within that amount of time.
+
+## Security and Authorization
```
To prevent a replay attack (hah, as if that would happen over https), should we validate the timestamp is fewer than `n` seconds old? Is there a `n` such that Discord guarantees the request will arrive within said `n` seconds?

msciotti replied:
@night would probably be a better answer for that. I think that in doing the signature validation, you should be guaranteed that it's fine, but I'm not the security expert 😄

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+For those developers looking to make more organized and complex groups of commands, look no further than subcommands and groups.
+
+**Subcommands** organize your command by **specifying actions within a command or group**.
```
```suggestion
**Subcommands** organize your commands by **specifying actions within a command or subcommand group**.
```

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+**Subcommands** organize your command by **specifying actions within a command or group**.
+
+**Subcommand Groups** organize your **subcommands** by **grouping commands by similar action or resource within a command**.
```
```suggestion
**Subcommand Groups** organize your **subcommands** by **grouping subcommands by similar action or resource within a command**.
```

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+These are not enforced rules. You are free to use subcommands and groups however you'd like; it's just how we think about them.
+
+> danger
+> Using subcommands or subcommand groups will make your base command unusable. You can't access the base `/permissions` if you also have `/permissions add | remove` as subcommands or subcommand groups
```
Given it is stated that the UI might change so the subcommands are on a single command, will this still hold true.

msciotti replied:
Technically yes but maybe the wording isn't the best. If you have `/permissions add | remove`, then `/permissions` by itself will never be a valid command. I'll think about the wording.

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+### Example Walkthrough
+
+Let's look at an example. Let's imagine you run a moderation bot. You want to make a `/permissions` command that can do the following:
```
Let's imagine you want to know whether a user has a role called "Muted" and you don't have the id. Whoops, you cannot do that with the current system.

^ Just me trying to remind you that whenever you add smarter tokens (if smarter tokens means more perms) to also give roles or something. Given that Discord is pushing commands to be stateless, this is a requirement (especially since you might not have a bot in a guild to keep track of the roles).

msciotti replied:
Well, a few things:

> you cannot do that with the current system.

If you still need permissions within the guild, you should still be using a bot user. I even explicitly stated that at the start of the docs.

> add smarter tokens (if smarter tokens means more perms) to also give roles or something. 

When I offhandedly mentioned smarter tokens, we're actually thinking more than this. What if the interaction token was a short-lived token scoped to the _user's_ permissions? So if I want to, say, create a channel via a bot, the bot wouldn't have to worry about its own permissions _and_ checking the user. It could just take the interactions token and hit the API and get a response without needing any permissions of its own.

^ that's all blue sky right now but an idea :)

Past this, let's talk in feature requests and comments off this PR tomorrow.

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+![A command with no arguments. It says /permissions](command.png)
+
+Now we have a command named `permissions`. We want this command to be able to affect users roles. Rather than making two separate commands, we can use subcommand groups. We want to use subcommand groups here because we are grouping commands on a similar resource: `user` or `role`.
```
```suggestion
Now we have a command named `permissions`. We want this command to be able to affect users and roles. Rather than making two separate commands, we can use subcommand groups. We want to use subcommand groups here because we are grouping commands on a similar resource: `user` or `role`.
```

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+You'll notice that a command like this **will not show up** in the command explorer. That's because groups are effectively "folders" for commands, and we've made two empty folders. So let's continue.
+
+After we scope our command to the `user` or `role` groups, we want to be able to either `get` or `edit` their permissions. Within the subcommand groups, we can make subcommands for `get` and `edit`:
```
```suggestion
After we scope our subcommand to the `user` or `role` groups, we want to be able to either `get` or `edit` their permissions. Within the subcommand groups, we can make subcommands for `get` and `edit`:
```

I'm not 100% sure though, given you say "scope our command" originally, which might mean it functions as a type of choice.

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+                        },
+                        {
+                            "name": "channel",
+                            "description": "The channel permissions to edit",
```
Maybe the `description` should mention that if it is left out, the guild permissions will be displayed?

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+Fetch all of the global commands for your application. Returns an array of [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/applicationcommand) objects.
+
+## Create Global Application Command % POST /applications/{application.id#DOCS_TOPICS_OAUTH2/application-object}/commands
```
Is there a limit to the payload size for this endpoint? (Hopefully it's large enough, the previous permission example was 1kb minified 😰)

msciotti replied:
A limit to the payload size for creating commands?

A5rocks replied:
Like how there's limits to embeds and whatnot, I guess. I was mostly just asking for the sake of asking, though.

msciotti replied:
Those limits are mostly just because our `send_message` has a max filesize on it I believe

Snazzah commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+| Field | Type          | Description         |
+|-------|---------------|---------------------|
+| name  | string        | name of the choice  |
+| value | string or int | value of the choice |
```
Is there any limit to either `name` or `value` in this case?

msciotti replied:
Name is 1-100 chars, good catch

Snazzah commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+| tts?              | bool             | is the response TTS                                                                         |
+| content           | string           | message content                                                                             |
+| embeds?           | array of embeds  | supports up to 10 embeds                                                                    |
+| allowed_mentions? | allowed mentions | [allowed mentions](#DOCS_RESOURCES_CHANNEL/allowed-mentions-object) object                  |
```
From what i've seen, `flags` would be here too.

msciotti replied:
I've intentionally left `flags` off for now because ephemeral messages aren't quite done (UI-wise) on all platforms.

Snazzah reviewed (changes requested):
Docs are missing out on talking about message flags or using the `EPHEMERAL` flag.

Snazzah commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+## Edit Followup Message % PATCH /webhooks/application.id/{interaction.token#DOCS_INTERACTIONS_SLASH_COMMANDS/interaction}/messages/{message.id#DOCS_RESOURCES_CHANNEL/message-object}
+
+Edits a followup message for an Interaction. Functions the same as [Edit Webhook Message](#DOCS_RESOURCES_WEBHOOK/edit-webhook-message).
```
Does this do the same behavior as setting `wait` to `true` in executing commands?

(meant to put this in *creating* a follow-up message)

msciotti replied:
Not entirely sure what you mean. Do you mean will it return the message object when creating a followup message?

Snazzah replied:
Yeah, that was my question. Meant to say executing webhooks not commands.

msciotti committed:
PR feedback

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+| data?\*    | ApplicationCommandInteractionData | the command data payload                                |
+| guild_id   | snowflake                         | the guild it was sent from                              |
+| channel_id | snowflake                         | the channel it was sent from                            |
+| member     | GuildMember                       | guild member data for the invoking user                 |
```
Is this guaranteed to have the `User` attached, unlike `CREATE_MESSAGE`?

msciotti replied:
Yes the user object is guaranteed

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+| guild_id   | snowflake                         | the guild it was sent from                              |
+| channel_id | snowflake                         | the channel it was sent from                            |
+| member     | GuildMember                       | guild member data for the invoking user                 |
+| token      | string                            | a contintuation token for responding to the interaction |
```
```suggestion
| token      | string                            | a continuation token for responding to the interaction |
```
Wording for the description is a little weird, but it's fine.

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+After receiving an interaction, you must respond to acknowledge it. This may be a `pong` for a `ping`, a message, or simply an acknowledgement that you have received it and will handle the command async.
+
+Interaction responses may choose to "eat" the user's command input if you do not wish to have their slash command show up as message in chat. This may be helpful for slash commands, or commands whose responses are asynchronous or ephemeral messages.
```
I feel "swallow" is the more widely used term here, but "eat" is a fun word.

msciotti replied:
Yknow, I _knew_ "eat" wasn't quite it but I knew it was some sort of eating-based word. I'm keeping it now though I've become attached :p

msciotti committed:
add s

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+###### InteractionResponseType
+
+| Name                     | Value | Description                                                       |
+|--------------------------|-------|-------------------------------------------------------------------|
+| Pong                     | 1     | ACK a `Ping`                                                      |
+| Acknowledge              | 2     | ACK a command without sending a message, eating the user's input  |
+| ChannelMessage           | 3     | respond with a message, eating the user's input                   |
+| ChannelMessageWithSource | 4     | respond with a message, showing the user's input                  |
+| ACKWithSource            | 5     | ACK a command without sending a message, showing the user's input |
```
For both `2` and `5`, if you provide `data` a message still gets sent, which is against the behaviour described here.

Also I would have found a bitflag more intuitive but too late to change that I guess. (`Flag.DisplaySource | Flag.SendMessage` would be `4`, for example. I did this in my repo "slash-worker")

msciotti replied:
That sounds like a bug, we can investigate.

A5rocks commented on docs/topics/Community_Resources.md:
```diff
@@ -39,6 +39,13 @@ Many of these libraries are represented in the [unofficial, community-driven Dis
 
+## Interactions
+
+[Interactions and Slash Commands](#DOCS_INTERACTIONS_SLASH_COMMANDS/) are the great, new way of making a Discord bot. The following open-source libraries provide help for the security and authentication checks that are mandatory if you are receiving Interactions via outgoing webhook. They also include some types for the Interactions data models.
```
The comma after `great` feels weird.

Snazzah commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+
+## Create Followup Message % POST /webhooks/application.id/{interaction.token#DOCS_INTERACTIONS_SLASH_COMMANDS/interaction}
+
+Create a followup message for an Interaction. Functions the same as [Execute Webhook](#DOCS_RESOURCES_WEBHOOK/execute-webhook)
```
Can this use different interaction types (like `ChannelMessageWithSource`) or is it completely restricted to the parameters from webhooks? If it is restricted, will we be able to use `file` in interaction responses or would it only be here?

advaith1 replied:
Followups are only webhook messages, types are only for the initial response

msciotti replied:
Advaith is correct that only the original response supports the different types since any followup message is just like sending a normal message.

I don't believe we support `file` yet, but we should. Had to make some tradeoffs in the architecture to start but definitely noted :)

Fabricio20 commented on docs/interactions/Slash_Commands.md:
```diff
@@ -0,0 +1,841 @@
+| INTEGER           | 4     |
+| BOOLEAN           | 5     |
+| USER              | 6     |
+| CHANNEL           | 7     |
```
Is type 7 here any channel, only text channels or voice channels as well?

ThaTiemsz replied:
Any channel - from text channels to channel categories.

msciotti replied:
Any channel within the guild, yes. So no DM channels but everything else including categories, store channels, etc.

bsian03 commented on docs/topics/OAuth2.md:
```diff
@@ -386,6 +387,7 @@ Returns the bot's OAuth2 application info.
 | primary_sku_id?        | snowflake                                                  | if this application is a game sold on Discord, this field will be the id of the "Game SKU" that is created, if exists     |
 | slug?                  | string                                                     | if this application is a game sold on Discord, this field will be the URL slug that links to the store page               |
 | cover_image?           | string                                                     | if this application is a game sold on Discord, this field will be the hash of the image on store embeds                   |
+| flags                  | int                                                        | the application's public flags                                                                                            |
```
Is there any specific reason why this isn't called `public_flags`?

msciotti replied:
I think that all the flags on an application actually are public, as opposed to users where we specifically differentiate. Calling them the public flags covers my butt for now but I can confirm.

erkinalp replied:
It would be useful for future-proofing for a future where bots could be boosted.

msciotti committed:
spelling
